### PR TITLE
Add support for Rust and Ruby programming languages

### PR DESCRIPTION
**Pull request by michaelalper-cezik-mw**
Imported Ruby & Rust tree-sitter grammars, added an extensive source file for both languages and adjusted nodeTypesConfig accordingly to assure basic support.

### CODE REVIEW

**Review by mylinhdao — APPROVED** [review]
LGTM

File: src/parser/config/nodeTypesConfig.json [hunk 1]
```diff
@@ -1948,10 +1948,10 @@
     },
     {
         "expression": "field_declaration_list",
-        "metrics": ["classes"],
```
Comment (ce-bo) [inline src/parser/config/nodeTypesConfig.json:1951]: Was this a wrong mapping? If yes, would you please add a regression test? At least one test should have failed in my opinion.

File: src/parser/config/nodeTypesConfig.json [hunk 2]
```diff
@@ -7036,5 +7036,1433 @@
...
+    {
+        "expression": "singleton_class",
+        "metrics": [],
+        "type": "statement",
+        "category": "",
+        "languages": ["rb"]
+    },
+    {
+        "expression": "singleton_method",
+        "metrics": [],
```
Comment (ce-bo) [inline src/parser/config/nodeTypesConfig.json:7441]: Are these also relevant for functions metric?

File: src/parser/config/nodeTypesConfig.json (hunk 1 above)
Comment (ResistantBear) [reply src/parser/config/nodeTypesConfig.json:1951]: If I remember correctly, the issue was that this did work for C++, but not for Rust. As `class_specifier` and `struct_specifier` should be the more straight forward choice for C++ anyway, we decided to count them instead for C++.

Comment (ResistantBear) [reply src/parser/config/nodeTypesConfig.json:1951]: I do not know about Go, as there were no test cases provided by you for the classes metric for that language while mentioning the reason that there are no classes in Go. But I see now that we actually counted this as class for Go, mh. We might look into that again.

File: test/parser/RustMetrics.test.ts [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    describe("parses Rust real lines of code metric", () => {
+        it("should count correctly for a non-empty file, ignoring comments and empty lines", () => {
+            testFileMetric(
+                rsTestResourcesPath + "rust-example-code.rs",
+                FileMetric.realLinesOfCode,
+                70
+            );
+        });
+    });
+});
```
Comment (ResistantBear) [inline test/parser/RustMetrics.test.ts:61]: There is no tests for the functions metric.

File: test/parser/RubyMetrics.test.ts [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    describe("parses Ruby real lines of code metric", () => {
+        it("should count correctly for a non-empty file, ignoring comments and empty lines", () => {
+            testFileMetric(
+                rbTestResourcesPath + "ruby-example-code.rb",
+                FileMetric.realLinesOfCode,
+                62
+            );
+        });
+    });
+});
```
Comment (ResistantBear) [inline test/parser/RubyMetrics.test.ts:57]: There is no tests for the functions metric.

File: src/parser/config/nodeTypesConfig.json (hunk 2 above)
Comment (ResistantBear) [reply src/parser/config/nodeTypesConfig.json]: I have checked. We should indeed count them. Accordingly, I have added a mapping to functions and complexity as well as test cases.

File: src/parser/config/nodeTypesConfig.json [hunk 5]
```diff
@@ -7036,5 +7036,1433 @@
...
+    {
+        "expression": "retry",
+        "metrics": [],
+        "type": "statement",
+        "category": "",
+        "languages": ["rb"]
+    },
+    {
+        "expression": "singleton_class",
+        "metrics": [],
```
[the remark this answers is not in the document]
Comment (ResistantBear) [reply src/parser/config/nodeTypesConfig.json:7434]: It seems like this gets generated if we have a `class << self` construct in Ruby. I have only seen examples for this inside of a class definition for defining something like static methods to the surrounding class. In this case, not counting this token yields the expected result, as adding static methods to a class is no new class definition.

File: test/parser/RubyMetrics.test.ts (hunk 4 above)
Comment (ResistantBear) [reply test/parser/RubyMetrics.test.ts:75]: I have added that myself now.

**Review by ResistantBear — CHANGES_REQUESTED** [review]
Adding the max_nesting_level metric is blocked until #179 is discussed and implemented.

**Review by ResistantBear — APPROVED** [review]

**Pull request merged**